### ISSUE TITLE
On-demand archiving of temporary files produced by scrapers

### ISSUE THREAD

**Issue by benoit74** [openzim/zimfarm#913]
The reported issue is that it is very difficult to debug `zimit` issues on the Zimfarm because the WARC files that are produced by Browsertrix crawler are lost once the recipe is finished. 

This causes:
- long time and resources lost crawling again the same website
- uncertainties about the fact that the WARC files we use in local tests are really the same than the ones which caused the issue on the Zimfarm

On a normal basis, we do not need / want to store these WARC files, just like we do not store gutenberg temporary DB and other build assets. But it would be great that we could do it when we know there is a high probability that we will need the WARC files (e.g. when debugging Zimit 2).

As discussed live in our last weekly dev meeting, there are basically three places where this could be implemented:
- in warc2zim
- in zimit
- in zimfarm

Since the WARCs are produced by Browsertrix crawler / Zimit, it feels odd to implement this in warc2zim since it would mean the scraper is archiving its input files which are not modified at all by the scraper.

Implementing this in `zimit` is probably more straightforward: we do not take the risk to modify shared code base and it is quickly deployed on all workers

Implementing this in `zimfarm` is not expected to be very difficult (we already have an upload "function" for logs and ZIMs) and it will be valuable for other scrapers as well. But since it will be done in the worker, it will be more "complex" to deploy and might have a bigger impact in case of issue.

I'm now quite convinced that `zimfarm` is the right place to implement this.

WDYT?

**Comment by benoit74** [comment]
After discussing with Renaud, we came to the following conclusions.

- Zimfarm is the proper place to implement the feature
- We should implement a real feature capable to archive on-demand any set of files produced by the scraper ; this should not be the normal behavior but could help debug many scrapers in fact
- The selection of which files to archive will be done by a new configuration item at the recipe level, displayed in the UI, specifying the glob expression to use to select files to archive
- The archiving process will be automatic, at the end of the task, just before cleanup
- Files to archive will be compressed in a single file for convenient retrieval (at the expense of temporarily doubling the storage used on the scraper FS => feature definitely needs to be used with caution ; we might consider to delete files on-the-go as we are close to the task termination anyway, hence additionally consuming only the size of the biggest file to archive)
- Archive will be pushed to the same target (S3 bucket) than logs (and hence benefit from same expiration policy)
- A new button will be displayed to download the archive when available

Nota: updating the Zimfarm worker is not a complex task because we will update the task-worker which is launched by the worker-manager for every new task and constantly checked for new versions, deployment is hence mostly immediate

**Comment by rgaudin** [comment]
> Files to archive will be compressed in a single file for convenient retrieval

I think an uncompressed archive might be better suited but there are pros and cons.

**Comment by kelson42** [comment]
Sournds good to me.

> * Files to archive will be compressed in a single file for convenient retrieval (at the expense of temporarily doubling the storage used on the scraper FS => feature definitely needs to be used with caution ; we might consider to delete files on-the-go as we are close to the task termination anyway, hence additionally consuming only the size of the biggest file to archive)

Let's try tar and uploading straight?!

**Comment by benoit74** [comment]
I love the "tar-on-the-fly" approach, but the proposed implementation is very dirty (running a subprocess to upload a file ... 😫) and we will have to check if it works as well on Wasabi. At least we know it might work.

**Comment by benoit74** [comment]
Maybe [S3fs](https://pypi.org/project/s3fs/) is worth considering, it seems to be maintained and used by a relevant share of users given the issues activity.

**Comment by mgautierfr** [comment]
I cannot resist myself to take my other hat and, as its creator, propose the [arx format](https://github.com/jubako/arx). It automatically compress compressible file while keeping other uncompressed. Archive are `mountable` so it is not necessary to extract the whole archive to read it.

Release 0.2 is in the pipe and streamable creation is something I thinking about since a time and I think it could be doable (not for 0.2)

(You can forget this message if you want, the subliminal message effect is enough for me)

**Comment by benoit74** [comment]
I had a more detailed look into Zimfarm codebase.

For ZIM and log upload, we use almost the same approach:
- an environment variable specifies where files must be uploaded ; SFTP, SCP and S3 targets are supported
  - we have two variables, one for ZIMs and one for logs
- an upload container is created for each file to upload
  - e.g. when we have 10 ZIMs to upload, we start the upload container 10 times
- when needed (docker logs), the file is created in advance on the filesystem 

This means that using a "tar-on-the-fly" approach is quite disruptive, because:
- we never processed whole folders / set of files before
- we won't be able to support all kinds of target (SFTP / SCP / S3) for this new feature

I suggest to stick to the original idea:
- the task manager will create a tar (uncompressed) of files matching a given glob pattern
- to save space, the task manager will delete files as they are added to the tar archive
- the upload container will upload this tar file to a target

I suggest to name this new thing an `archive`, i.e. we will upload `ZIMs`, `logs` and `archives`.

We will hence have:
- a new environment variable in Zimfarm to specify archive target location
- for openzim, a new S3 bucket (we could decide to have different retention rules than logs)

**Comment by rgaudin** [comment]
LGTM
